src/core_modules/capture-core/components/WidgetReferral/useReferral.js

```diff
@@ -1,48 +1,50 @@
 import { userStores } from '../../storageControllers/stores';
 
+const getRelationshipTypeFromIndexedDB = () => {
+    const storageController = getUserStorageController();
+    return storageController.getAll(userStores.RELATIONSHIP_TYPES);
+};
+
 export const useReferral = (programStageId: string) => {
     const [currentReferralStatus, setReferralStatus] = useState();
-    const [selectedRelationshipTypes, setSelectedRelationshipTypes] = useState([]);
 
-    const getRelationshipTypeFromIndexedDB = useCallback(async () => {
-        const storageController = getUserStorageController();
-        const allRelationshipTypes = await storageController.getAll(userStores.RELATIONSHIP_TYPES);
-        const referrableRelationshipTypes = allRelationshipTypes
-            .filter(relationshipType => relationshipType.referral && relationshipType.access.data.write);
-
-        setSelectedRelationshipTypes(referrableRelationshipTypes);
-    }, []);
+    const { data: relationshipTypes } = useQuery('relationshipTypes', () => getRelationshipTypeFromIndexedDB(), {
+        select: allRelationshipTypes => allRelationshipTypes
+            ?.filter(relationshipType => relationshipType.referral && relationshipType.access.data.write) ?? [],
+    });
+    const selectedRelationshipType = currentReferralStatus === referralStatus.REFERRABLE ?
+        relationshipTypes?.[0] : undefined;
+    const constraint = selectedRelationshipType?.toConstraint?.programStage?.id === programStageId ?
+        selectedRelationshipType?.fromConstraint : selectedRelationshipType?.toConstraint;
 
     useEffect(() => {
-        getRelationshipTypeFromIndexedDB();
-    }, [getRelationshipTypeFromIndexedDB]);
-
-    useEffect(() => {
-        if (selectedRelationshipTypes.length === 1) {
-            const { fromConstraint, toConstraint } = selectedRelationshipTypes[0];
-
-            if ((fromConstraint?.programStage?.id === programStageId
-                || toConstraint?.programStage?.id === programStageId)
-            ) {
-                setReferralStatus(referralStatus.REFERRABLE);
+        if (relationshipTypes) {
+            if (relationshipTypes.length === 1) {
+                const { fromConstraint, toConstraint } = relationshipTypes[0];
+
+                if ((fromConstraint?.programStage?.id === programStageId
+                    || toConstraint?.programStage?.id === programStageId)
+                ) {
+                    setReferralStatus(referralStatus.REFERRABLE);
+                }
+            } else if (relationshipTypes.length > 1) {
+                setReferralStatus(referralStatus.AMBIGUOUS_REFERRALS);
             }
-        } else if (selectedRelationshipTypes.length > 1) {
-            setReferralStatus(referralStatus.AMBIGUOUS_REFERRALS);
         }
     }, [
-        setSelectedRelationshipTypes,
-        selectedRelationshipTypes,
+        relationshipTypes,
         setReferralStatus,
         programStageId,
     ]);
```

Comment:
I know you didn't write this from scratch, but can we refactor this to use useMemo (ie, get rid of useEffect and the useState)? I don't see any side-effects here.